substrate/primitives/runtime/src/proving_trie/base16.rs

```diff
@@ -0,0 +1,327 @@
+	fn query(&self, key: &Key) -> Option<Value> {
+		let trie = TrieDBBuilder::new(&self.db, &self.root).build();
+		key.using_encoded(|s| trie.get(s))
+			.ok()?
+			.and_then(|raw| Value::decode(&mut &*raw).ok())
+	}
+
+	/// Create a compact merkle proof needed to prove a single key and its value are in the trie.
+	///
+	/// This function makes a proof with latest substrate trie format (`LayoutV1`), and is not
+	/// compatible with `LayoutV0`.
+	fn create_proof(&self, key: &Key) -> Result<Vec<u8>, DispatchError> {
+		sp_trie::generate_trie_proof::<LayoutV1<Hashing>, _, _, _>(
+			&self.db,
+			self.root,
+			&[key.encode()],
+		)
+		.map_err(|err| TrieError::from(*err).into())
+		.map(|structured_proof| structured_proof.encode())
+	}
+
+	/// Verify the existence of `key` and `value` in a given trie root and proof.
+	///
+	/// Proofs must be created with latest substrate trie format (`LayoutV1`).
+	fn verify_proof(
+		root: &Hashing::Out,
+		proof: &[u8],
+		key: &Key,
+		value: &Value,
+	) -> Result<(), DispatchError> {
+		verify_proof::<Hashing, Key, Value>(root, proof, key, value)
+	}
+}
+
+impl<Hashing, Key, Value> ProofToHashes for BasicProvingTrie<Hashing, Key, Value>
+where
+	Hashing: sp_core::Hasher,
+	Hashing::Out: MaxEncodedLen,
+{
+	// This base 16 trie does not directly expose the depth of the trie, so we can roughly calculate
+	// it assuming the data in the proof are hashes, and the number of hashes present will tell us
+	// the depth of the trie.
+	fn proof_to_hashes(proof: &[u8]) -> Result<u32, DispatchError> {
+		let proof_size = proof.len() as u32;
+		let hash_len = Hashing::Out::max_encoded_len() as u32;
+		// A base 16 trie is expected to include the data for 15 hashes per layer.
+		let layer_len = hash_len.checked_mul(15).ok_or(ArithmeticError::Overflow)?;
+		let proof_size_round_up = proof_size
+			.checked_add(layer_len)
+			.ok_or(ArithmeticError::Overflow)?
+			.checked_sub(1)
+			.ok_or(ArithmeticError::Underflow)?;
+
+		let depth = proof_size_round_up
+			.checked_div(layer_len)
+			.ok_or(ArithmeticError::DivisionByZero)?;
+		Ok(depth)
+	}
+}
+
+/// Verify the existence of `key` and `value` in a given trie root and proof.
+///
+/// Proofs must be created with latest substrate trie format (`LayoutV1`).
+pub fn verify_proof<Hashing, Key, Value>(
+	root: &Hashing::Out,
+	proof: &[u8],
+	key: &Key,
+	value: &Value,
+) -> Result<(), DispatchError>
+where
+	Hashing: sp_core::Hasher,
+	Key: Encode,
+	Value: Encode,
+{
+	let structured_proof: Vec<Vec<u8>> =
+		Decode::decode(&mut &proof[..]).map_err(|_| TrieError::DecodeError)?;
+	sp_trie::verify_trie_proof::<LayoutV1<Hashing>, _, _, _>(
+		&root,
+		&structured_proof,
+		&[(key.encode(), Some(value.encode()))],
+	)
+	.map_err(|err| TrieError::from(err).into())
+}
+
+/// Verify the existence of multiple `items` in a given trie root and proof.
+///
+/// Proofs must be created with latest substrate trie format (`LayoutV1`).
```

Comment:
Same, can be V0, I think only things that call `generate_for` require this V1 V0 distinction.